File: trunk/Source/Runtime/CSharp/Halak.Bibim.Runtime/GameModuleNode.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;

namespace Halak.Bibim
{
    public sealed class GameModuleNode
    {
        #region Fields
        private GameModule module;
        private GameModuleNode parent;
        private List<GameModuleNode> children;
        private GameModuleTree tree;
        #endregion

        #region Properties
        public GameModule Module
        {
            get { return module; }
        }

        public GameModuleNode Parent
        {
            get { return parent; }
        }

        public IEnumerable<GameModuleNode> Children
        {
            get { return children; }
        }

        public GameModuleTree Tree
        {
            get { return tree; }
            internal set
            {
                tree = value;
            }
        }

        public bool IsRoot
        {
            get { return parent != null && tree != null; }
        }
        #endregion

        #region Constructors
        public GameModuleNode()
        {
        }

        public GameModuleNode(GameModule module)
        {
            this.module = module;
        }

        public GameModuleNode(GameModule module, int childrenCapacity)
            : this(module)
        {
            children = new List<GameModuleNode>(childrenCapacity);
        }
        #endregion

        #region Methods
        #region Children Manipulation
        public GameModuleNode AttachChild(GameModule item)
        {
            if (item.Node == null)
                item.Node = new GameModuleNode(item);

            AttachChild(item.Node);

            return item.Node;
        }

        public void AttachChild(GameModuleNode item)
        {
            if (item.parent == this)
            {
                Debug.Assert(children != null && children.Contains(item));
                return;
            }

            bool treeChanged = (item.parent == null || item.parent.tree != tree);

            GameModuleNode oldParent = item.parent;
            if (oldParent != null)
            {
                // Detach를 거치지 않고 직접 제거합니다.
                item.parent.children.Remove(item);
                item.parent = null;

                // GameStructure.OnNodeDetached의 명세대로 GameStructure가 바뀌었을 때만 호출합니다.
                if (oldParent.tree != null && treeChanged)
                    oldParent.tree.OnNodeDetached(oldParent, item);
            }

            Debug.Assert(item.parent == null);

            item.parent = this;
            item.tree = tree;
            if (children == null)
                children = new List<GameModuleNode>();
            children.Add(item);

            // GameStructure.OnNodeAttached의 명세대로 GameStructure가 바뀌었을 때만 호출합니다.
            if (tree != null && treeChanged)
                tree.OnNodeAttached(this, item);
        }

        public bool DetachChild(GameModuleNode item)
        {
            if (children == null || item.parent != this || item.tree != tree)
                return false;

            bool removed = children.Remove(item);
            Debug.Assert(removed); // 이미 item의 부모가 this인지 확인했기 때문에 반드시 있어야합니다.

            item.parent = null;

            if (tree != null)
                tree.OnNodeDetached(this, item);

            return true;
        }

        public void DetachAllChildren()
        {
            if (children == null)
                return;

            List<GameModuleNode> temporaryChildren = children;

            if (tree != null)
            {
                for (int i = temporaryChildren.Count - 1; i >= 0; i--)
                {
                    GameModuleNode item = temporaryChildren[i];
                    tree.OnNodeDetached(this, item);
                    item.parent = null;
                }
            }
            else
            {
                for (int i = temporaryChildren.Count - 1; i >= 0; i--)
                    temporaryChildren[i].parent = null;
            }
        }
        #endregion

        #region Find
        public GameModule FindChild(uint id)
        {
            GameModuleNode node = FindChildNode(id);
            if (node != null)
                return node.Module;
            else
                return null;
        }

        public GameModule FindChildByClassID(uint classID)
        {
            GameModuleNode node = FindChildNodeByClassID(classID);
            if (node != null)
                return node.Module;
            else
                return null;
        }

        public GameModuleNode FindChildNode(uint id)
        {
            if (id == GameModule.UnspecifiedID)
                return null;

            foreach (GameModuleNode item in children)
            {
                if (item.Module != null && item.Module.ID == id)
                    return item;

            }

            foreach (GameModuleNode item in children)
            {
                GameModuleNode foundNode = FindChildNode(id);
                if (foundNode != null)
                    return foundNode;
            }

            return null;
        }

        public GameModuleNode FindChildNodeByClassID(uint classID)
        {
            foreach (GameModuleNode item in children)
            {
                if (item.Module != null && item.Module.ClassID == classID)
                    return item;
            }

            foreach (GameModuleNode item in children)
            {
                GameModuleNode foundNode = FindChildNodeByClassID(classID);
                if (foundNode != null)
                    return foundNode;
            }

            return null;
        }

        public GameModuleNode FindChildNode(GameModule item)
        {
            // 원래는 모든 자식 Node들을 순회하며 검사해야하지만,
            // GameModule에서 자신을 가진 Node를 아는 관계로 최적화할 수 있었습니다.
            // GameModule의 조상Node중 this가 있는지 확인합니다.

            if (item.Node == null || item.Node.tree != tree)
                return null;

            if (IsAncestor(item.Node, this))
                return item.Node;
            else
                return null;
        }

        private static bool IsAncestor(GameModuleNode descendant, GameModuleNode ancestor)
        {
            if (descendant.parent == null)
                return false;
            if (descendant.parent == ancestor)
                return true;
            else
                return IsAncestor(descendant.parent, ancestor);
        }
        #endregion

        #region Internal
        internal void OnModuleAliveChanged()
        {
            // Alive => 상위 GameModule와 하위 GameModule들을 모두 Alive로 만듭니다.
            // Dead  => 하위 GameModule들을 모두 Dead상태로 만듭니다.

            bool alive = module.Alive;

            if (alive && parent != null)
                parent.module.Alive = true;
            foreach (GameModuleNode item in children)
                item.module.Alive = alive;
        }
        #endregion
        #endregion
    }
}

File: trunk/src/Module/GameModuleNode.cpp
#include <Bibim/Config.h>
#include <Bibim/GameModuleNode.h>
#include <Bibim/Assert.h>
#include <Bibim/GameModule.h>
#include <Bibim/GameModuleTree.h>
#include <algorithm>

namespace Bibim
{
    GameModuleNode::GameModuleNode()
        : module(nullptr),
          parent(nullptr),
          tree(nullptr)
    {
    }

    GameModuleNode::GameModuleNode(GameModule* module)
        : module(module),
          parent(nullptr),
          tree(nullptr)
    {
        BBAssert(module);
    }

    GameModuleNode::GameModuleNode(GameModule* module, int childrenCapacity)
        : module(module),
          parent(nullptr),
          tree(nullptr)
    {
        BBAssert(module);
        children.reserve(childrenCapacity);
    }

    GameModuleNode::~GameModuleNode()
    {
        DetachAllChildren();

        if (module)
        {
            GameModule* temporaryModule = nullptr;
            std::swap(temporaryModule, module);
     
            temporaryModule->node = nullptr;
            delete temporaryModule;
        }
    }

    GameModuleNode* GameModuleNode::AttachChild(GameModule* item)
    {
        BBAssert(item);

        if (item->node == nullptr)
            item->node = new GameModuleNode(item);

        AttachChild(item->node);

        return item->node;
    }

    void GameModuleNode::AttachChild(GameModuleNode* item)
    {
        BBAssert(item);

        if (item->parent == this)
        {
            BBAssertDebug(std::find(children.begin(), children.end(), item) != children.end());
            return;
        }

        const bool treeChanged = (item->parent == nullptr || item->parent->tree != tree);

        if (GameModuleNode* oldParent = item->parent)
        {
            // Detach를 거치지 않고 직접 제거합니다.
            NodeCollection::iterator it = std::find(oldParent->children.begin(), oldParent->children.end(), item);
            BBAssertDebug(it != oldParent->children.end());
            oldParent->children.erase(it);
            item->parent = nullptr;

            // GameModuleTree::OnNodeDetached의 명세대로 GameModuleTree가 바뀌었을 때만 호출합니다.
            if (oldParent->tree && treeChanged)
                oldParent->tree->OnNodeDetached(oldParent, item);
        }

        BBAssertDebug(item->parent == nullptr);

        item->parent = this;
        item->tree = tree;
        children.push_back(item);

        // GameModuleTree::OnNodeAttached의 명세대로 GameModuleTree가 바뀌었을 때만 호출합니다.
        if (tree && treeChanged)
            tree->OnNodeAttached(this, item);
    }

    bool GameModuleNode::DetachChild(GameModuleNode* item)
    {
        BBAssert(item);

        if (item->parent != this || item->tree != this->tree)
            return false;

        NodeCollection::iterator it = std::find(children.begin(), children.end(), item);
        BBAssertDebug(it != children.end()); // 이미 item의 부모가 this인지 확인했기 때문에 반드시 있어야합니다.
        children.erase(it);

        item->parent = nullptr;

        if (tree)
            tree->OnNodeDetached(this, item);

        delete item;

        return true;
    }

    void GameModuleNode::DetachAllChildren()
    {
        NodeCollection temporaryChildren;
        temporaryChildren.swap(children);

        if (tree)
        {
            for (NodeCollection::reverse_iterator it = temporaryChildren.rbegin(); it != temporaryChildren.rend(); ++it)
            {
                tree->OnNodeDetached(this, *it);
                delete (*it);
            }
        }
        else
        {
            for (NodeCollection::reverse_iterator it = temporaryChildren.rbegin(); it != temporaryChildren.rend(); ++it)
                delete (*it);
        }
    }

    GameModule* GameModuleNode::FindChild(int id) const
    {
        if (GameModuleNode* foundNode = FindChildNode(id))
            return foundNode->GetModule();
        else
            return nullptr;
    }

    GameModule* GameModuleNode::FindChildByClassID(int id) const
    {
        if (GameModuleNode* foundNode = FindChildNodeByClassID(id))
            return foundNode->GetModule();
        else
            return nullptr;
    }

    GameModuleNode* GameModuleNode::FindChildNode(int id) const
    {
        if (id == GameModule::UnspecifiedID)
            return nullptr;

        for (NodeCollection::const_iterator it = children.begin(); it != children.end(); ++it)
        {
            if ((*it)->GetModule() && (*it)->GetModule()->GetID() == id)
                return (*it);
        }

        for (NodeCollection::const_iterator it = children.begin(); it != children.end(); ++it)
        {
            if (GameModuleNode* foundNode = (*it)->FindChildNode(id))
                return foundNode;
        }

        return nullptr;
    }

    GameModuleNode* GameModuleNode::FindChildNodeByClassID(int id) const
    {
        for (NodeCollection::const_iterator it = children.begin(); it != children.end(); ++it)
        {
            if ((*it)->GetModule() && (*it)->GetModule()->GetClassID() == id)
                return (*it);
        }

        for (NodeCollection::const_iterator it = children.begin(); it != children.end(); ++it)
        {
            if (GameModuleNode* foundNode = (*it)->FindChildNodeByClassID(id))
                return foundNode;
        }

        return nullptr;
    }

    GameModuleNode* GameModuleNode::FindChildNode(GameModule* item) const
    {
        // 원래는 모든 자식 Node들을 순회하며 검사해야하지만,
        // GameModule에서 자신을 가진 Node를 아는 관계로 최적화할 수 있었습니다.
        // GameModule의 조상Node중 this가 있는지 확인합니다.

        if (item->node == nullptr || item->node->tree != tree)
            return nullptr;

        struct ThisIsAncestor
        {
            static bool Do(GameModuleNode* descendant, const GameModuleNode* ancestor)
            {
                if (descendant->parent == nullptr)
                    return false;

                if (descendant->parent == ancestor)
                    return true;
                else
                    return Do(descendant->parent, ancestor);
            }
        };

        if (ThisIsAncestor::Do(item->node, this))
            return item->node;
        else
            return nullptr;
    }

    void GameModuleNode::OnModuleDestructed()
    {
        module = nullptr;
    }

    void GameModuleNode::OnModuleAliveChanged()
    {
        // Alive => 상위 GameModule와 하위 GameModule들을 모두 Alive로 만듭니다.
        // Dead  => 하위 GameModule들을 모두 Dead상태로 만듭니다.

        const bool alive = module->GetAlive();

        if (alive && parent)
            parent->module->SetAlive(true);
        for (NodeCollection::iterator it = children.begin(); it != children.end(); ++it)
            (*it)->module->SetAlive(alive);
    }
}